quinn-proto/src/connection.rs

```diff
@@ -778,51 +778,62 @@ impl Connection {
-                if !self.state.is_drained() {
-                    debug!(self.log, "got stateless reset");
-                    mux.timer_stop(Timer::LossDetection);
-                    mux.timer_stop(Timer::Close);
-                    mux.timer_stop(Timer::Idle);
-                    mux.emit(ConnectionError::Reset.into());
-                    self.state = State::Drained;
-                }
-                return;
-            }
-        }
-
         trace!(self.log, "connection got packet"; "len" => packet.payload.len());
         let was_handshake = self.is_handshaking();
         let was_closed = self.state.is_closed();
 
+        let mut stateless_reset = false;
+        if let Some(token) = self.params.stateless_reset_token {
+            stateless_reset =
+                packet.payload.len() >= 16 && packet.payload[packet.payload.len() - 16..] == token;
+        }
+
         let result = match self.decrypt_packet(was_handshake, &mut packet, crypto_update) {
             Err(Some(e)) => {
                 warn!(self.log, "got illegal packet"; "reason" => %e);
                 Err(e.into())
             }
             Err(None) => {
                 debug!(self.log, "failed to authenticate packet");
-                return;
+                if stateless_reset {
+                    Err(ConnectionError::Reset)
+                } else {
+                    return;
+                }
             }
             Ok(number) => {
-                if let Some(number) = number {
-                    if self.dedup.insert(number) {
+                let duplicate =
+                    number.and_then(|n| if self.dedup.insert(n) { Some(n) } else { None });
+
+                if let Some(number) = duplicate {
+                    if stateless_reset {
+                        Err(ConnectionError::Reset)
+                    } else {
                         warn!(
                             self.log,
                             "discarding possible duplicate packet {packet}",
                             packet = number
                         );
                         return;
                     }
+                } else {
+                    if !was_closed {
+                        self.on_packet_authenticated(mux, now, number);
+                    }
+                    self.handle_connected_inner(mux, now, number, packet)
                 }
-                if !was_closed {
-                    self.on_packet_authenticated(mux, now, number);
-                }
-                self.handle_connected_inner(mux, now, number, packet)
             }
         };
 
+        if stateless_reset && !self.state.is_drained() {
+            if let Err(ConnectionError::Reset) = result {
+                debug!(self.log, "got stateless reset");
+                mux.timer_stop(Timer::LossDetection);
+                mux.timer_stop(Timer::Close);
+                mux.timer_stop(Timer::Idle);
+            }
+        }
```

Comment:
Is checking the value of `stateless_reset` here redundant to the pattern-match? I think we could also move the body of this into the match statement below, retaining the `!is_drained()` guard.